protocols/virtualdefund/virtualdefund.go

```diff
@@ -61,17 +73,105 @@ func newObjective(preApprove bool, vFixed state.FixedPart, initialOutcome outcom
 		status = protocols.Unapproved
 	}
 
+	V, found := getChannel(request.ChannelId)
+	if !found {
+		return Objective{}, fmt.Errorf("could not find channel %s", request.ChannelId)
+	}
+
+	initialOutcome := V.PostFundState().Outcome[0]
+
+	// This logic assumes a single hop virtual channel.
+	// Currently this is the only type of virtual channel supported.
+	alice := V.Participants[0]
+	intermediary := V.Participants[1]
+	bob := V.Participants[2]
+
+	var toMyLeft, toMyRight *consensus_channel.ConsensusChannel
+	var ok bool
+
+	switch request.MyAddress {
+	case alice:
+		toMyRight, ok = getConsensusChannel(intermediary)
+		if !ok {
+			return Objective{}, fmt.Errorf("could not find a ledger channel between %v and %v", alice, intermediary)
+		}
+	case intermediary:
+		toMyLeft, ok = getConsensusChannel(alice)
+		if !ok {
+			return Objective{}, fmt.Errorf("could not find a ledger channel between %v and %v", alice, intermediary)
+		}
+		toMyRight, ok = getConsensusChannel(bob)
+		if !ok {
+			return Objective{}, fmt.Errorf("could not find a ledger channel between %v and %v", intermediary, bob)
+		}
+	case bob:
+		toMyLeft, ok = getConsensusChannel(intermediary)
+		if !ok {
+			return Objective{}, fmt.Errorf("could not find a ledger channel between %v and %v", intermediary, bob)
+		}
+	default:
+		return Objective{}, fmt.Errorf("client address not found in an expected participant index")
+
+	}
 	return Objective{
 		Status:         status,
 		InitialOutcome: initialOutcome,
-		PaidToBob:      paidToBob,
-		VFixed:         vFixed,
+		PaidToBob:      request.PaidToBob,
+		VFixed:         V.FixedPart,
 		Signatures:     [3]state.Signature{},
-		MyRole:         myRole,
+		MyRole:         V.MyIndex,
 		ToMyLeft:       toMyLeft,
 		ToMyRight:      toMyRight,
+	}, nil
+
+}
+
+// ConstructObjectiveFromState takes in a message and constructs an objective from it.
+// It accepts the message, myAddress, and a function to to retrieve ledgers from a store.
+func ConstructObjectiveFromState(
+	initialState state.State,
+	myAddress types.Address,
+	getChannel GetChannelByIdFunction,
+	getTwoPartyConsensusLedger GetTwoPartyConsensusLedgerFunction,
+) (Objective, error) {
+	channelId, err := initialState.ChannelId()
+	if err != nil {
+		return Objective{}, err
 	}
 
+	paidToBob, err := calculatePaidToBob(initialState, getChannel)
```

Comment:
Should we document somewhere the following assumption (and tag it with a TODO):

> Because payments are not currently implemented, we are not able to query or verify how much was paid. So the current behaviour is as follows:
> * whomever calls `CloseVirtualChannel` gets to set `paidToBob` however they like
> * the responder will go along with it

(if I have got this right!)

Reply:
Maybe we should also add something about payments being "owned" by the consuming application, rather than `go-nitro` directly?

Relevant discussion: https://github.com/statechannels/go-nitro/discussions/617